public/app/plugins/datasource/cloud-monitoring/components/VisualMetricQueryEditor.tsx

```diff
@@ -38,12 +40,24 @@ function Editor({
   const [labels, setLabels] = useState<{ [k: string]: any }>({});
   const { projectName, groupBys, crossSeriesReducer } = query;
   const metricType = getMetricType(query);
+  const [timeState, setTime] = useState<{ from: string; to: string } | null>(null);
+  const timeSrv = getTimeSrv();
+
+  const useTime = (time: { from: string; to: string }) => {
+    if (timeState !== null && (timeState.from !== time.from || timeState.to !== time.to)) {
+      setTime({ ...timeSrv.time });
+    } else if (!timeState) {
+      setTime({ ...timeSrv.time });
+    }
+  };
+
+  useTime(timeSrv.time);
 
   useEffect(() => {
     if (projectName && metricType) {
       datasource.getLabels(metricType, refId, projectName).then((labels) => setLabels(labels));
     }
-  }, [datasource, groupBys, metricType, projectName, refId, crossSeriesReducer]);
+  }, [datasource, groupBys, metricType, projectName, refId, crossSeriesReducer, timeState]);
```

Comment:
this is a bit brittle, if you remove `timeState` here, nothing will tell you that you are breaking this. I see that `getLabels` is using `this.timeSrv.timeRange()`. Why don't you store the `TimeRange` in the state instead and pass it as an argument? That way you are sure that you are calling `getLabels` with the right range.

Reply:
not using `groupBys` and `crossSeriesReducer` already sounds like a bug and I don't know if that's on purpose :sweat_smile: 

> There isn't a need to pass TimeRange as an arg as the value used within the function would be the correct one

yes, but it feels more natural to calculate it here since this component is the one in charge of re-calling the function in case the TimeRange here. You are calculating the `time` here just to detect changes and recalculate the `range`. Isn't it simpler to just calculate the range and store that?

Reply:
Also a good point. I'm now considering doing a bigger refactor of the `VisualMetricQueryEditor` because I'm not entirely sure why this component is separate to the `Metrics` component when there isn't much in that component specifically, wdyt?

Reply:
yep, makes sense!